server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlVisitorHelper.java

```diff
@@ -150,4 +158,65 @@ public static Position getAdjustedEndPosition(
     }
     return end;
   }
+
+  public static ExecSqlWheneverNode.WheneverConditionType getConditionType(Db2SqlParser.Dbs_wheneverContext ctx) {
+    ParserRuleContext ruleContext = ((ParserRuleContext) ctx);
+    ExecSqlWheneverNode.WheneverConditionType conditionType = ExecSqlWheneverNode.WheneverConditionType.NOT_FOUND;
+
+
+    if (ruleContext.getChildCount() >= 3) {
+      ParseTree pt = ruleContext.getChild(1);
+      String value = pt.getText().trim().toUpperCase();
+
+      if (Objects.equals(value, "SQLERROR")) {
+        conditionType = ExecSqlWheneverNode.WheneverConditionType.SQLERROR;
+      } else if (Objects.equals(value, "SQLWARNING")) {
+        conditionType = ExecSqlWheneverNode.WheneverConditionType.SQLWARNING;
+      }
+    }
+    return conditionType;
+  }
+
+  public static Pair<ExecSqlWheneverNode.WheneverType, String> getWheneverType(Db2SqlParser.Dbs_wheneverContext ctx) {
+    ParserRuleContext ruleContext = ((ParserRuleContext) ctx);
```

Comment:
How could that possibly be more code? The whole function is basically this
```
    if (ctx.CONTINUE() != null)
      return Pair.of(ExecSqlWheneverNode.WheneverType.CONTINUE, null);
    return Optional.ofNullable(ctx.dbs_host_name_container())
        .map(c -> c.dbs_host_names())
        .filter(hn -> !hn.isEmpty())
        .map(c -> c.get(0))
        .map(ParseTree::getText)
        .map(t -> Pair.of(ExecSqlWheneverNode.WheneverType.GOTO, t))
        .orElse(/* whatever is appropriate to return when there was a parser error */);

```

Reply:
Also for GO TO you need to take .map(c -> c.get(1)), not .map(c -> c.get(0))

Reply:
DO is not part of the grammar, that is why I was asking where does DO come from.
GOTO/GO TO is handled by the parser.
https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/blob/af0a479250ad0fbcaaf101e28df2b15add27b74d/server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlParser.g4#L1150

Reply:
The `get(0)` gets the first `dbs_host_names` context, not the next token.

Reply:
Wow, got it. We need to add DO to the grammar

Reply:
I have mixed up IBM i and Enterprise COBOL